### PR TITLE
ECDSA - signature conversion and refactor

### PR DESCRIPTION
**Pull request by mwoods-figure**
- Introduce value classes for `ASN1Signature` and `BTCSignature`

- Add decode() methods on ECDSASignature for ASN1/BTC signatures

- Decode an ASN.1-encoded array of bytes into an `(r,s)` pair of integers.

- Add more comments

### CODE REVIEW

File: signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+/**
+ * Bitcoin-specific signature format operations.
+ */
+object BTC {
+    /**
+     * [encode] returns the ECDSA signature as a ByteArray of `r || s`, * where both `r` and `s` are encoded into 32
+     * byte big endian integers.
+     */
+    fun encode(r: BigInteger, s: BigInteger, curve: Curve): BTCSignature {
```
Comment (jphorec) [inline signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:47]: I have a small nitpick, does it make sense to call this a bitcoin signature? Isn't the R and S pair just a specific encoding of an ECDSA signature and not necessarily bitcoin specific? Not sure if there could be a better naming convention

**Review by jphorec — APPROVED** [review]

File: signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt (hunk 1 above)
Comment (mtps) [reply signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:47]: Yea, they are technically called `canonical` ecdsa signatures (it's mentioned all over in go's [ecdsa](https://pkg.go.dev/github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa#section-readme))

File: signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import org.bouncycastle.asn1.ASN1Sequence
+import org.bouncycastle.asn1.DERSequenceGenerator
+
+@JvmInline
+value class BTCSignature private constructor(private val value: ByteArray) {
+    companion object {
+        fun fromByteArray(bytes: ByteArray): BTCSignature = BTCSignature(bytes)
+    }
+
+    fun toByteArray(): ByteArray = value
```
Comment (mtps) [inline signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:19]: Should have a `toBigIntegerPair()` as well to mimic asn1 below?

**Review by mtps — DISMISSED** [review]

File: signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            ASN1Signature.fromByteArray(it.toByteArray())
+        }
+
+    /**
+     * Decode an ASN.1 signature into an `(r,s)` integer pair.
+     *
+     * @param bytes The signature bytes to decode.
+     * @return The decoded `(r,s)` integer pair as a [BigIntegerPair].
+     */
+    fun decode(bytes: ByteArray): BigIntegerPair =
```
Comment (mtps) [inline signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:138]: future note: If BigIntegerPair is converted to a data class, callers can use decomposition to pull the r and s values directly as `val (r, s) = ASN1.decode(bytes)`

File: signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        require(rBytes.size <= 32) { "cannot encode r into BTC Format, size overflow (${rBytes.size} > 32)" }
+        require(sBytes.size <= 32) { "cannot encode s into BTC Format, size overflow (${sBytes.size} > 32)" }
+
+        val signature = ByteArray(64)
+        // 0 pad the byte arrays from the left if they aren't big enough.
+        System.arraycopy(rBytes, 0, signature, 32 - rBytes.size, rBytes.size)
+        System.arraycopy(sBytes, 0, signature, 64 - sBytes.size, sBytes.size)
+
+        return BTCSignature.fromByteArray(signature)
+    }
```
Comment (mtps) [inline signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:83]: Sibling object has both encode and decode, @mwoods-figure can we get the decode method pulled in here? I think that would be a little more consistent.

Comment (mwoods-figure) [reply signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:83]: @mtps Sure!

File: signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt (hunk 3 above)
Comment (mwoods-figure) [reply signer/src/main/kotlin/io/provenance/hdwallet/signer/SignatureFormat.kt:138]: That's a great point

**Review by jphorec — APPROVED** [review]

**Pull request merged**